python-client/tests/communications/test_dto_parsing.py

```diff
@@ -0,0 +1,414 @@
+    ],
+)
+def test_func_argument_missing_fields(func_argument_dict):
+    with pytest.raises(ValidationError):
+        websocket.FuncArgument.parse_obj(func_argument_dict)
+
+
+@pytest.mark.parametrize(
+    "func_argument_dict",
+    [
+        {"name": TEST_FUNC_ARGUMENT_NAME, "none": True},  # None argument
+        {
+            "name": TEST_FUNC_ARGUMENT_NAME,
+            "none": False,
+            "model": {"project_key": TEST_PROJECT_KEY, "id": TEST_MODEL_ID},
+        },  # Model argument
+        {
+            "name": TEST_FUNC_ARGUMENT_NAME,
+            "none": False,
+            "dataset": {"project_key": TEST_PROJECT_KEY, "id": TEST_DATASET_ID, "sample": False},
+        },  # Dataset argument
+        {
+            "name": TEST_FUNC_ARGUMENT_NAME,
+            "none": False,
+            "float": TEST_FUNC_ARGUMENT_FLOAT,
+        },  # float argument
+        {
+            "name": TEST_FUNC_ARGUMENT_NAME,
+            "none": False,
+            "int": TEST_FUNC_ARGUMENT_INT,
+        },  # int argument
+        {
+            "name": TEST_FUNC_ARGUMENT_NAME,
+            "none": False,
+            "str": TEST_FUNC_ARGUMENT_STR,
+        },  # str argument
+        {
+            "name": TEST_FUNC_ARGUMENT_NAME,
+            "none": False,
+            "bool": TEST_FUNC_ARGUMENT_BOOL,
+        },  # bool argument
+        {
+            "name": TEST_FUNC_ARGUMENT_NAME,
+            "none": False,
+            "slicingFunction": {"project_key": TEST_PROJECT_KEY, "id": TEST_FUNC_ARGUMENT_SLICING_FUNCTION_ID},
+        },  # slicing function argument
+        {
+            "name": TEST_FUNC_ARGUMENT_NAME,
+            "none": False,
+            "transformationFunction": {
+                "project_key": TEST_PROJECT_KEY,
+                "id": TEST_FUNC_ARGUMENT_TRANSFORMATION_FUNCTION_ID,
+            },
+        },  # transformation function argument
+        {
+            "name": TEST_FUNC_ARGUMENT_NAME,
+            "none": False,
+            "kwargs": f"kwargs['{TEST_FUNC_ARGUMENT_NAME}'] = {TEST_FUNC_ARGUMENT_INT}",
+        },  # kwargs function argument
+        {
+            "name": TEST_FUNC_ARGUMENT_NAME,
+            "none": False,
+            "args": [
+                {
+                    "name": TEST_FUNC_ARGUMENTS_NAME1,
+                    "none": False,
+                    "int": TEST_FUNC_ARGUMENT_INT,
+                },  # int argument
+                {
+                    "name": TEST_FUNC_ARGUMENTS_NAME2,
+                    "none": False,
+                    "str": TEST_FUNC_ARGUMENT_STR,
+                },  # float argument
+            ],
+        },  # args function argument
+    ],
+)
+def test_func_argument(func_argument_dict):
+    func_argument = websocket.FuncArgument.parse_obj(func_argument_dict)
+    assert isinstance(func_argument, websocket.FuncArgument)
+    if not func_argument.is_none:
+        if func_argument.model is not None:
```

Comment:
I'm not sure doing one big test with parametrize and handling all cases is the best way to do that
It's making it quite difficult to read

Else, you could have an expected result dict, and use it ?

```python
 EXPECTED_RESULT = {
 "model": {
    "project_key"  : TEST_PROJECT_KEY,
    "model": { "id" :  TEST_MODEL_ID}
   },
 "args": [{ "is_none":False, "name": ....]
   
  ...
 }

def recursive_equality(obj, expected):
  if isinstance(obj, dict):
    assert all([hasattr(expected, k) for k in obj.keys()])
    for k, v in obj.items():
      recursive_equality(v, getattr(expected, k)
)

  elif isinstance(v, list):
    assert len(v) == len(expected)
    for elt_expected, elt in zip(v, expected) :
        recursive_equality(elt_expected, elt)

  else:
    assert obj == expected


 def test_func_argument(func_argument_dict):
    func_argument = websocket.FuncArgument.parse_obj(func_argument_dict)
   for k, expected in EXPECTED_RESULT:
      to_test = getattr(func_argument, k)
      if to_test  is not None:
        assert isinstance(to_test, websocket.ArtifactRef)
        recursive_equality(to_test, expected)
        return 
   assert False # Should not happen
   
```